src/twisted/internet/_signals.py

```diff
@@ -80,6 +89,169 @@ def isDefaultHandler():
     return signal.getsignal(signal.SIGCHLD) == signal.SIG_DFL
 
 
+class SignalHandling(Protocol):
+    """
+    The L{SignalHandling} protocol enables customizable signal-handling
+    behaviors for reactors.
+
+    A value that conforms to L{SignalHandling} has install and uninstall hooks
+    that are called by a reactor at the correct times to have the (typically)
+    process-global effects necessary for dealing with signals.
+    """
+
+    def install(self) -> None:
```

Comment:
I'm not sure if it's useful to look at it that way - but with the mess in the current implementation of `_WithSignalHandling` it's a bit hard to tell.

However, I think about the actual type signature of this method as `IO ()`.  Its purpose in life is to have a side-effect.  So, maybe you want that side-effect to happen twice because something else happened in between and the second call should really just re-install everything?

Reply:
And as detailed in one of my other comments it turns out something, probably trial, manages to drive the process such that `.install(); ... .install()` really happens.

Reply:
https://github.com/twisted/twisted/issues/11763 and linked in the code